player/screenshot.c

```diff
@@ -127,6 +127,28 @@ static void append_filename(char **s, const char *f)
     talloc_free(append);
 }
 
+// Screenshot base filenames (especially with %{sub-text}) can be longer than
+// filesystems support (255 generally). Truncate it to fit with an extension.
+// If truncation produces an invalid UTF-8 codepoint, then chop that off.
+static void truncate_long_base_filename(char *s, const size_t ext_len)
+{
+    const size_t max_utf8_bytes = 255 - (ext_len + 1); // ext_len+1 for '.'
```

Comment:
255 is a magic number. It should not be hardcoded. It should _probably_ be MAX_PATH.

On windows MAX_PATH can be more than 255, because it has enough space to hold UTF8 of 260 `wchar_t` elements.

Also, if `extlen` is 255 or more, then `max_utf8_bytes` will wrap around to a huge number...

Reply:
It should not be `MAX_PATH` because `MAX_PATH` is not the basename length. `MAX_PATH` is to hold drive-letter + `":\"` + basename of 255 `wchar_t`/`char` elements + NUL terminator.

Per [Maximum Path Length Limitation](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation) for Windows:
> This type of path is composed of components separated by backslashes, each up to the value returned in the lpMaximumComponentLength parameter of the [GetVolumeInformation](https://learn.microsoft.com/en-us/windows/desktop/api/FileAPI/nf-fileapi-getvolumeinformationa) function (this value is commonly 255 characters).

Bothering to check `GetVolumeInformation` isn't worth doing though.
All relevant filesystem use 255 for segments of filename (including for non-Windows OSes).

If `ext_len` is 255 let mpv blow up because that's an absurd case to care about.

Reply:
> It should not be `MAX_PATH`

Which is why I said _probably_, I.e. you should figure whether it should be MAX_PATH or something else, like MAX_NAME.

The point is that 255 should not be hardcoded. It should be appropriate for the current platform, and if it's not MAX_PATH and not MAX_NAME then you should figure out what it needs to be, without calling APIs.

It should probably be some existing constant of the platform, and not hardcoded inside this function.

> If `ext_len` is 255 let mpv blow up because that's an absurd case to care about.

In your applications maybe. Not in mpv.

You mean that the example you gave which should be fixed are not absurd cases, and so we should really care about them, like this?
> `screenshot-template="a🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂"`

Luckily for you though, it won't blow up, but it will also not work.

It's not rocket science. Please fix it correctly.

Reply:
> Which is why I said probably, I.e. you should figure whether it should be MAX_PATH or something else, like MAX_NAME.

Which is what I did and why 255.

> The point is that 255 should not be hardcoded. It should be appropriate for the current platform, and if it's not MAX_PATH and not MAX_NAME then you should figure out what it needs to be, without calling APIs.

Don't hardcode but also figure it out without calling APIs? Okay...


>> If ext_len is 255 let mpv blow up because that's an absurd case to care about.

> In your applications maybe. Not in mpv.

The extension comes from mpv. If mpv decides to use 255 character long extensions then that is mpv's fault.

>> screenshot-template="a🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂🌂"

> Luckily for you though, it won't blow up, but it will also not work.
It's not rocket science. Please fix it correctly.

You're missing something if you think that won't work, or why it was listed as an example for testing removal of invalidated UTF-8 codepoints due to truncation.

Reply:
> It should probably be some existing constant of the platform

I don't know if it's available.

We have this:
https://github.com/mpv-player/mpv/blob/ef4c6dfcf5d2b9a5d96a4cd2c6deaf49bbbe38c4/osdep/io.c#L539-L547

But it's only used privately at this C file when enumerating files in a directory.

Not sure how to solve this in general. I don't think we shold change the global MAX_PATH either.

Reply:
`NAME_MAX` seems to available in `limits.h` for Linux/macos as 255. It's also in my mingw64/msys2's `limits.h` but behind a `_POSIX_` ifdef. BSDs have `MAXNAMELEN` which is 255 as far as I can tell.

Something like this, hardcoding 255, or calling out to `GetVolumeInformation`/`pathconf(_PC_NAME_MAX)`
```
#include <limits.h>
#ifndef NAME_MAX
 #ifdef MAXNAMELEN
  #define NAME_MAX MAXNAMELEN
 #else
  #define NAME_MAX 255
 #endif
#endif
```

Reply:
For win32 you can use `_MAX_FNAME` https://learn.microsoft.com/en-us/cpp/c-runtime-library/path-field-limits

Note that `_MAX_FNAME` includes space for terminating null, while `NAME_MAX` does not.

Reply:
> Note that `_MAX_FNAME` includes space for terminating null, while `NAME_MAX` does not.

I didn't include it because of that but yeah a Windows specific define could be `(_MAX_FNAME-1)` which is 255